File: Data Structures & Algorithms/count-number-of-islands/submission-3.py
class Solution:
    def dfs(self, i, j, grid, m, n):
        grid[i][j] = "#"
        nr = [0, -1, 1, 0]
        nc = [-1, 0, 0, 1]
        for k in range(4):
            dr, dc = i + nr[k], j + nc[k]
            if 0 <= dr < m and 0 <= dc < n and grid[dr][dc] == "1":
                self.dfs(dr, dc, grid, m, n)
        
        
    def numIslands(self, grid: List[List[str]]) -> int:
        m = len(grid)
        n = len(grid[0])
        count = 0
        for i in range(0,m):
            for j in range(0,n):
                if grid[i][j] == "1":
                    self.dfs(i, j, grid, m, n)
                    count += 1
        return count

File: Data Structures & Algorithms/valid-tree/submission-3.cpp
class Solution {
public:
    int count = 0;

    void dfs(int node, vector<vector<int>>& adj, vector<bool>& visited) {
        visited[node] = true;
        count++;

        for (int neighbor : adj[node]) {
            if (!visited[neighbor]) {
                dfs(neighbor, adj, visited);
            }
        }
    }

    bool validTree(int n, vector<vector<int>>& edges) {
        if (edges.size() != n - 1)
            return false;

        vector<vector<int>> adj(n);

        for (auto edge : edges) {
            int a = edge[0];
            int b = edge[1];

            adj[a].push_back(b);
            adj[b].push_back(a);
        }

        vector<bool> visited(n, false);

        dfs(0, adj, visited);

        return count == n;
    }
};
